### PR TITLE
Added bluetooth Logic

### PR DESCRIPTION
**Pull request by jeverd**
Added Bluetooth logic. The Home screen UI is just temporary - used for testing. This will change once @khalidr12 code is merged. 

RN-Flipper was removed because it prevents gRPC streams events to come through on Android. Flipper is used mainly for debugging, so don't see a problem removing it.

### CODE REVIEW

File: src/services/lasnService.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+const {
+  SubscribeRequest,
+  LinkRequest,
+  DisconnectRequest,
+} = require('../../protos/lasn/v1/lasn_pb');
+
+export const getLasnStream = (lasnService, userId) => {
```
Comment (lnogueir) [inline src/services/lasnService.js:7]: Wondering if we could call it subscription rather than stream.
```suggestion
export const getLasnSubscription = (lasnService, userId) => {
```

**Review by lnogueir — COMMENTED** [review]
I know this is just draft, but just 2 things I wanted to mention in case I forget in the future.

File: android/app/build.gradle [hunk 2]
```diff
@@ -268,18 +268,18 @@ dependencies {
 
     implementation "androidx.swiperefreshlayout:swiperefreshlayout:1.0.0"
 
-    debugImplementation("com.facebook.flipper:flipper:${FLIPPER_VERSION}") {
-        exclude group:'com.facebook.fbjni'
-    }
```
Comment (jeverd) [inline android/app/build.gradle:273]: grpc streams and flipper don't work well together, so had to comment/remove flipper from android. Should not be a problem since its only used for debugging, which we dont use anyways

File: src/components/screens/Home/HomeSearchingView.js [hunk 3]
```diff
@@ -3,17 +3,20 @@ import {ActivityIndicator, StyleSheet, View} from 'react-native';
...
     <View style={styles.container}>
       <ActivityIndicator size="large" color={colors.black} />
-      <ApproachText style={styles.searchingText}>Searching</ApproachText>
-      <ApproachText style={styles.searchingSmallText}>
-        Make sure you are close enough to a person. On average, the Bluetooth
-        connection has a maximum radius of 15 meters
+      <ApproachText style={styles.loadingText}>
+        {renderLoadingText()}
       </ApproachText>
+      <ApproachText style={styles.loadingSmallText}>Please wait!</ApproachText>
```
Comment (jeverd) [inline src/components/screens/Home/HomeSearchingView.js:19]: probably should change this, just kept is as placeholder

File: src/services/lasnService.js (hunk 1 above)
Comment (jeverd) [reply src/services/lasnService.js]: tbh, I think is stream better bc we are listening for events(stream of events type of thing). However im down to change it to subscription, lmk.

File: src/services/bluetoothService.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import {BleManager} from 'react-native-ble-plx';
+import BLEAdvertiser from 'react-native-ble-advertiser';
+import {Platform} from 'react-native';
+
+// IOS TO ANDROID BLE ADVERTISING: https://github.com/vitorpamplona/react-native-ble-advertiser/issues/39?fbclid=IwAR0Gd9tbQe04HOARjVp6ibc6iQDeAW0tctbJlL_CzBhUr8X4CJ5y-QtbslA
+const COMPANY_ID = 0x4c;
+const SCAN_MANUF_DATA = Platform.OS === 'android' ? [] : [1, 0];
+const SERVICE_ID_SUFFIX = 'ae12';
```
[the remark this answers is not in the document]
Comment (jeverd) [reply src/services/bluetoothService.js]: good call, added.

File: src/services/bluetoothService.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  for (const serviceId of serviceIds) {
+    if (serviceId.endsWith(SERVICE_ID_SUFFIX)) {
+      return serviceId;
+    }
+  }
+  return;
+};
+
+export const startBluetoothScanningService = (bleManager, fp) => {
+  bleManager.startDeviceScan(null, {allowDuplicates: true}, fp);
```
Comment (jeverd) [inline src/services/bluetoothService.js:33]: `{allowDuplicates: true}` needed for IOS to allow for unlinking timeout to occur

File: src/components/screens/Home/HomeLoadingView.js [hunk 6]
```diff
@@ -3,17 +3,14 @@ import {ActivityIndicator, StyleSheet, View} from 'react-native';
...
   return (
     <View style={styles.container}>
       <ActivityIndicator size="large" color={colors.black} />
-      <ApproachText style={styles.searchingText}>Searching</ApproachText>
-      <ApproachText style={styles.searchingSmallText}>
-        Make sure you are close enough to a person. On average, the Bluetooth
-        connection has a maximum radius of 15 meters
-      </ApproachText>
+      <ApproachText style={styles.loadingText}>Loading...</ApproachText>
+      <ApproachText style={styles.loadingSmallText}>Please wait!</ApproachText>
```
Comment (lnogueir) [inline src/components/screens/Home/HomeLoadingView.js:13]: Tbh, just get rid of this.

File: src/sagas/bluetoothSaga.js [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+  }
+}
+
+export function* getInitialBluetoothState(bleManager) {
+  const bluetoothState = yield call(getBluetoothState, bleManager);
+  const bluetoothAction = new Action(
+    CHECK_DEVICE_BLUETOOTH_STATE,
+    bluetoothState,
+  );
+  yield put({...bluetoothAction});
```
Comment (lnogueir) [inline src/sagas/bluetoothSaga.js:76]: Isn't this equivalent?
```suggestion
  yield put(bluetoothAction);
```

File: src/sagas/bluetoothSaga.js [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+}
+
+function* bluetoothScannerChannelEventHandler(event) {
+  switch (event.type) {
+    case BLUETOOTH_SCANNER_EVENT_TYPES.BLUETOOTH_CHANNEL_SCANNED_DEVICED: {
+      const scannedDeviceAction = new Action(DEVICE_SCANNED, event.payload);
+      yield put({...scannedDeviceAction});
```
Comment (lnogueir) [inline src/sagas/bluetoothSaga.js:83]: same here. and all other `put`s below

File: src/sagas/bluetoothSaga.js [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    bluetoothScannerChannel = yield call(
+      createBluetoothScannerChannel,
+      bleManager,
+    );
+    while (true) {
+      const event = yield take(bluetoothScannerChannel);
+      yield call(bluetoothScannerChannelEventHandler, event);
+    }
+  } catch (e) {
+    console.log('[SAGA] bluetoothScanner error', e);
```
Comment (lnogueir) [inline src/sagas/bluetoothSaga.js:99]:
```suggestion
    console.error('[SAGA] bluetoothScanner error', e);
```
Same for the logs below.

File: src/sagas/bluetoothSaga.js [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+      deviceScanned: take(DEVICE_SCANNED),
+      timeout: delay(UNLINK_TIMEOUT),
+    });
+    if (!deviceScanned) {
+      // timeout has been reached.
+      // check to make sure we in conencted state
+      const networkState = yield select(selectNetworkConnectionState);
+      if (networkState === NETWORK_STATE.CONNECTED) {
+        // todo: Add unlinking grpc call
+        Alert.alert('Unlinking time!!!!');
```
Comment (lnogueir) [inline src/sagas/bluetoothSaga.js:162]: @SergioSanchez12 we need to add your unlinking code here.

File: src/sagas/bluetoothSaga.js [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+    // while (true) {
+    // yield take(BLUETOOTH_START_UP);
+    console.log('BLE STARTUP!!!');
+    let tasks = yield all([
+      fork(bluetoothSeqTask, bleAdvertiser, bleManager),
+      fork(unlinkTimeoutWatch),
+    ]);
+    const bluetoothSeqStartupSuccess = new Action(BLUETOOTH_START_UP_SUCCESS);
+    yield put({...bluetoothSeqStartupSuccess});
+    console.log('BLE SUCCESS!!!');
```
Comment (lnogueir) [inline src/sagas/bluetoothSaga.js:180]: Do we wanna keep these logs here

File: src/sagas/bluetoothSaga.js [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+      }
+    }
+  }
+}
+
+export function* bluetoothRootSaga(bleManager) {
+  const bleAdvertiser = yield call(createBleAdvertiser);
+  try {
+    // while (true) {
+    // yield take(BLUETOOTH_START_UP);
```
Comment (lnogueir) [inline src/sagas/bluetoothSaga.js:172]: should we delete these commented out lines?

File: src/sagas/bluetoothSaga.js [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+    console.log('bluetoothRootSaga starting teardown!');
+    yield call(cancelTasks, tasks);
+    yield all([
+      take(BLUETOOTH_BROADCASTER_TASK_STOPPED),
+      take(BLUETOOTH_SCANNER_TASK_STOPPED),
+    ]);
+    const teardownSuccessAction = new Action(BLUETOOTH_TEAR_DOWN_SUCCESS);
+    yield put({...teardownSuccessAction});
+    console.log('TEARING DOWN BLE!!!');
+    // }
```
Comment (lnogueir) [inline src/sagas/bluetoothSaga.js:192]: same here

File: src/sagas/lasnSaga.js [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+        filterSelfFromNearbyUsersList,
+        userId,
+        lasnEvent.payload,
+      );
+      const nearbyUserListAction = new Action(
+        UPDATE_NEARBY_USER_LIST,
+        nearbyUserList,
+      );
+      console.log('NEARBY USER LIST', nearbyUserList);
+      // console.log("PHOTO LIST", nearbyUserList[0].getPhotosList());
```
Comment (lnogueir) [inline src/sagas/lasnSaga.js:41]: should we get rid of these lines

File: src/sagas/lasnSaga.js [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+export function* lasnConnectionWatcher() {
+  try {
+    let lasnEventListenerTask, linkWatcherTask;
+    const lasnServiceClient = new LasnServicePromiseClient(GRPC_SERVICES_URL);
+    const userId = yield select(selectUserId);
+    // while (true) {
+    // startup
+    // yield take(LASN_START_UP);
```
Comment (lnogueir) [inline src/sagas/lasnSaga.js:98]: should we get rid of these comments

File: src/sagas/lasnSaga.js [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+    console.log('lasnWatcher: starting disconnecting');
+    yield call(disconnectFromLasn, lasnServiceClient, userId);
+    yield fork(cancelTasks, [linkWatcherTask, lasnEventListenerTask]);
+
+    yield take(LASN_EVENT_LISTENER_TASK_STOPPED);
+    console.log('cancelling tasks!!!');
+    console.log('WAITING FOR ACTION');
+    const tearDownSuccessAction = new Action(LASN_TEAR_DOWN_SUCCESS);
+    yield put({...tearDownSuccessAction});
+    // }
```
Comment (lnogueir) [inline src/sagas/lasnSaga.js:116]: same here

File: src/services/bluetoothService.js [hunk 17]
```diff
@@ -0,0 +1,51 @@
+import {BleManager} from 'react-native-ble-plx';
+import BLEAdvertiser from 'react-native-ble-advertiser';
+import {Platform} from 'react-native';
+import {BLE_SERVICE_SUFFIX} from '@env';
+
+// IOS TO ANDROID BLE ADVERTISING: https://github.com/vitorpamplona/react-native-ble-advertiser/issues/39?fbclid=IwAR0Gd9tbQe04HOARjVp6ibc6iQDeAW0tctbJlL_CzBhUr8X4CJ5y-QtbslA
+const COMPANY_ID = 0x4c;
```
Comment (lnogueir) [inline src/services/bluetoothService.js:7]: should this be something that could go on .env? or is this not specific to our app

File: src/services/bluetoothService.js [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    includeTxPowerLevel: false,
+  });
+};
+
+export const stopBluetoothScanning = bleManager => {
+  bleManager.stopDeviceScan();
+};
+
+export const stopBroadcastBluetoothSignal = bleAvertiser =>
+  bleAvertiser.stopBroadcast();
```
Comment (lnogueir) [inline src/services/bluetoothService.js:51]: can you fix the typo here: `bleAvertiser` -> `bleAdvertiser`

**Review by lnogueir — APPROVED** [review]
Insane job bro 👏🎉

LGTM other than the small nits.

**Comment by lnogueir** [comment]
I think before we merge we can actually add the code to unlink.

File: src/sagas/bluetoothSaga.js (hunk 10 above)
Comment (lnogueir) [reply src/sagas/bluetoothSaga.js:162]: Hoping we can add this code before merging this.

**Comment by lnogueir** [comment]
Also bro, if you could provide a brief explanation of how the Bluetooth works that'd be awesome. Maybe add the diagram to the description of this PR once you make it for the poster.

File: src/components/screens/Home/HomeUsersNearbyView.js [hunk 19]
```diff
@@ -49,7 +53,11 @@ export default function HomeUsersNearbyView({nearbyUsers}) {
 
   function NearbyUserCard(props) {
     const profileClickHandler = () => {
-      console.log('user profile tapped');
+      navigation.navigate(DMsScreenName, {
+        user2Id: props.nearbyUserId,
+        user2Name: props.name,
+        user2Image: props.photos[0],
```
Comment (lnogueir) [inline src/components/screens/Home/HomeUsersNearbyView.js:59]: we should make these changes on a separate PR.

File: src/components/screens/Home/HomeUsersNearbyView.js [hunk 20]
```diff
@@ -6,15 +6,18 @@ import {
   TouchableOpacity,
   View,
 } from 'react-native';
-import {useTheme} from '@react-navigation/native';
+import {useNavigation, useTheme} from '@react-navigation/native';
```
Comment (lnogueir) [inline src/components/screens/Home/HomeUsersNearbyView.js:9]: this should be imported on a separate PR.

File: src/actions/lasnActions.js [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+export const LASN_TEAR_DOWN_SUCCESS = 'LASN_TEAR_DOWN_SUCCESS';
+
+export const LASN_EVENT_LISTENER_TASK_STOPPED =
+  'LASN_EVENT_LISTENER_TASK_STOPPED';
+
+export const UPDATE_NETWORK_STATE = 'UPDATE_NETWORK_STATE';
+
+export const dispatchStartSearchingAction = dispatch => {
+  const action = new Action(START_SEARCHING);
+  dispatch({...action});
```
[the remark this answers is not in the document]
Comment (jeverd) [reply src/actions/lasnActions.js:24]: No - see error:

image

File: src/sagas/bluetoothSaga.js (hunk 11 above)
Comment (jeverd) [reply src/sagas/bluetoothSaga.js:180]: ya for now, make it easier to debug in case there are bugs. Will remove in another PR before production

File: src/actions/lasnActions.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (lnogueir) [reply src/actions/lasnActions.js:24]: lol wth. sounds good

File: src/channels/lasnChannel.js [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+      );
+    });
+
+    lasnStream.on('end', () => {
+      lasnEventHandler(LASN_EVENTS.CLOSE_CONNECTION);
+    });
+
+    const unsubscribe = () => {};
+
+    return unsubscribe;
```
[the remark this answers is not in the document]
Comment (jeverd) [reply src/channels/lasnChannel.js]: good catch, i forgot to add the closing of the stream and channel is closed. Added now.

File: src/sagas/bluetoothSaga.js (hunk 10 above)
Comment (jeverd) [reply src/sagas/bluetoothSaga.js]: Added

File: src/services/bluetoothService.js (hunk 17 above)
Comment (jeverd) [reply src/services/bluetoothService.js:7]: Usually yes, but in this case we set `COMPANY_ID ` to this to make IOS detect I believe android devices, so its not really specific to our app, will never change.  More info here: https://github.com/vitorpamplona/react-native-ble-advertiser/issues/39?fbclid=IwAR0Gd9tbQe04HOARjVp6ibc6iQDeAW0tctbJlL_CzBhUr8X4CJ5y-QtbslA

File: src/sagas/bluetoothSaga.js [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+export function* bluetoothBroadcaster(bleAdvertiser) {
+  try {
+    const deviceId = yield select(selectDeviceId);
+    while (true) {
+      yield call(broadcastBluetoothSignal, bleAdvertiser, deviceId);
+      yield delay(BROADCAST_FREQUENCY);
+      yield call(stopBroadcastBluetoothSignal, bleAdvertiser); // need to stop to be able to broadcast again.
+      //   console.log('BROADCAST RESULT', broadcastResult, stopBroadcastResult);
```
Comment (lnogueir) [inline src/sagas/bluetoothSaga.js:103]: Do we wanna remove this comment?

File: src/sagas/bluetoothSaga.js [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+  } finally {
+    console.log('[SAGA] Stopping: bluetoothSeqTask');
+  }
+}
+
+export function* bluetoothRootSaga(bleManager) {
+  const bleAdvertiser = yield call(createBleAdvertiser);
+  try {
+    console.log('BLE STARTUP!!!');
+    let tasks = yield all([fork(bluetoothSeqTask, bleAdvertiser, bleManager)]);
```
Comment (lnogueir) [inline src/sagas/bluetoothSaga.js:143]:
```suggestion
    const tasks = yield all([fork(bluetoothSeqTask, bleAdvertiser, bleManager)]);
```

File: src/sagas/index.js [hunk 25]
```diff
@@ -1 +1,97 @@
...
+  BLUETOOTH_START_UP_SUCCESS,
+  BLUETOOTH_TEAR_DOWN,
+  BLUETOOTH_TEAR_DOWN_SUCCESS,
+} from '../actions/bluetoothActions';
+import {Action} from '../actions';
+import {forkedDispatchActionWrapper} from './helperSaga';
+import {createBleManager} from '../services/bluetoothService';
+
+function* connectionStateWatcher() {
+  let networkStateAction = new Action(UPDATE_NETWORK_STATE);
```
Comment (lnogueir) [inline src/sagas/index.js:24]:
```suggestion
  const networkStateAction = new Action(UPDATE_NETWORK_STATE);
```

File: src/sagas/index.js [hunk 26]
```diff
@@ -1 +1,97 @@
...
+    networkStateAction.payload = NETWORK_STATE.DISCONNECTED;
+    yield fork(forkedDispatchActionWrapper, networkStateAction);
+  }
+}
+
+function* connectionHandler(bleManager) {
+  while (true) {
+    yield take(START_SEARCHING);
+    // start up bluetooth and lasn sagas
+    let [bluetoothTask, lasnTask] = yield all([
```
Comment (lnogueir) [inline src/sagas/index.js:60]:
```suggestion
    const [bluetoothTask, lasnTask] = yield all([
```

File: src/sagas/index.js [hunk 27]
```diff
@@ -1 +1,97 @@
...
+    yield take(LASN_TEAR_DOWN_SUCCESS);
+
+    yield all([cancel(bluetoothTask), cancel(lasnTask)]);
+    console.log('NETWORK SHUTDOWN!!!');
+  }
+}
+
+export function* rootSaga() {
+  yield take(REHYDRATE); // Wait for rehydrate to prevent sagas from running with empty store
+  let setNetworkAction = new Action(
```
Comment (lnogueir) [inline src/sagas/index.js:83]:
```suggestion
  const setNetworkAction = new Action(
```

**Review by lnogueir — COMMENTED** [review]
Nits: looks like we can use `const` in these places.

File: src/sagas/bluetoothSaga.js (hunk 23 above)
Comment (jeverd) [reply src/sagas/bluetoothSaga.js:103]: I occasionally uncomment this, so lets leave it in for now.
